doc/HOWTO-KAFKA.md

```diff
@@ -0,0 +1,226 @@
+
+In the Zookeeper shell:
+
+```
+   bin/zookeeper-server-start.sh config/zookeeper.properties
+```
+
+In the Server shell:
+
+```
+  bin/kafka-server-start.sh config/server.properties
+```
+
+In the Work shell, topics need to be added with `kafka-topics.sh` a la this, if you haven't done it
+before (or if you did it, but did not shut down the broker properly):
+
+```
+  bin/kafka-topics.sh --create --topic test-cluster.hpc.uio.no.sample --bootstrap-server localhost:9092
+```
+
+The topics to add are these (the last two are for control messages):
+
+```
+  test-cluster.hpc.uio.no.sample
+  test-cluster.hpc.uio.no.sysinfo
+  test-cluster.hpc.uio.no.job
+  test-cluster.hpc.uio.no.cluster
+  test-cluster.hpc.uio.no.control.node
+  test-cluster.hpc.uio.no.control.master
+```
+
+#### Running sonar and examining the data
+
+Then from the Sonar root directory, after building it, run Sonar in daemon mode with a suitable
+config file in the Sonar shell:
+
+```
+  target/debug/sonar daemon <config-file>
+```
+
+Sonar will run continuously and start pumping data to Kafka.
+
+For easy testing, run this in the Consumer shell to listen for `sysinfo` messages and echo them:
+
+```
+  bin/kafka-console-consumer.sh --topic 'test-cluster.hpc.uio.no.sysinfo' --bootstrap-server localhost:XXXX
+```
+
+Or run this in the Consumer shell to listen for `sample` messages and echo them:
+
+```
+  bin/kafka-console-consumer.sh --topic 'test-cluster.hpc.uio.no.sample' --bootstrap-server localhost:XXXX
+```
+
+#### Sending control messages
+
+(NOTE: Not currently implemented because there's no real need for it.)
+
+To send control messages to Sonar's compute node daemons:
+
+```
+  bin/kafka-console-producer.sh --bootstrap-server localhost:XXXX \
+      --topic test-cluster.hpc.uio.no.control.node \
+      --property parse.key=true
+```
+
+and then use TAB to separate key and value on each line.  A good test is `dump true` and
+`dump false`, but `exit` should work (without a value).
+
+#### Shutting down Kafka in an orderly way
+
+In the Work shell, in the Kafka root directory:
+
+```
+bin/kafka-server-stop.sh --bootstrap-server localhost:9092
+bin/zookeeper-server-stop.sh
+```
+
+#### Encryption
+
+To enable encryption, three things must happen:
+
+- generate and sign the necessary keys
+- update Kafka's `server.properties`
+- update the Sonar daemon's .ini file
+
+We will need a CA certificate (to be used by both Kafka and Sonar) and a key store containing the
+server's public and private keys (to be used by Kafka).  NOTE, the server keys will be tied to a
+particular server name.
+
+For testing, we will generate our own CA and key materials.  In util/ssl, there is a Makefile that
+will generate the necessary files: `sonar-ca.crt` is the CA certificate, and
+`sonar-kafka-keystore.pem` is the key store.  Just run `make all` to make the files for the local
+hostname.
+
+Having generated those, update Kafka's `server.properties` as follows, this is almost in the form of
```

Comment:
Clean this up.  Move the diffs from the last patch to this patch, and rewrite the prose here to reflect the availability of those files.  Keep the note that the paths may need to be updated.